### ISSUE TITLE
Can fastapi proxy another site as a response to the request?

### ISSUE THREAD

**Issue by hkhk366** [tiangolo/fastapi#1788]
This may be a silly question. Can fastapi proxy another site as a response to the request?

I did some research. All pages I found are Nginx reverse proxy fastapi. I already did that to solve the SSL problem.

My system workflow:
                                   

![image](https://user-images.githubusercontent.com/12952110/88448453-2beb6d00-ce0c-11ea-9578-331721107dbb.png)

RedirectResponse is a rewrite, not a proxy.
Static Files are for local files. 
HTMLResponse is for static HTML.

Can you give me some tips on how to proxy another site in fastapi? Thank you.

**Comment by eelkeh** [comment]
I have the same issue! Did you find a solution?

**Comment by mickeypash** [comment]
This is not quite working but it might be a start?
```
from fastapi import FastAPI, Response
import requests

app = FastAPI()

def _proxy(*args, **kwargs):
    t_resp = requests.request(
        method="GET",
        url="https://google.com/",
        allow_redirects=False)

    response = Response(content=t_resp.contents, status_code=t_resp.status_code)
    return response

@app.get("/")
def proxy():
    return _proxy()
```

**Comment by thomascrha** [comment]
How would I allow any path to be resolved by the proxy ? With flask you can do something like this 
```
from flask import Flask
from requests import get

app = Flask('__main__')
SITE_NAME = 'https://google.com/'

@app.route('/', defaults={'path': ''})
@app.route('/<path:path>')
def proxy(path):
  return get(f'{SITE_NAME}{path}').content

app.run(host='0.0.0.0', port=8080)
```

Can I allow wildcard paths with fastAPI ?

**Comment by eelkeh** [comment]
Any change that anyone made progress with this? I'm still looking for a solution

**Comment by jtbaker** [comment]
@eelkeh @thomas-maschler I just ran into this myself and found it was covered in the docs [here](https://fastapi.tiangolo.com/tutorial/path-params/#__code_13) referencing accessing a file system. A proxy should resolve the path same way.
```
from fastapi import FastAPI

app = FastAPI()


@app.get("/files/{file_path:path}")
async def read_file(file_path: str):
    return {"file_path": file_path}
```

**Comment by papb** [comment]
Hello @aebrahim and others, what if the request we want to proxy is streaming a lot of data as _input_? The solution by @aebrahim waits for all the request input to arrive first, and then will end up sending it all at once to the target server, am I right? If yes, do you have any idea on how to update the solution to better cover this case as well? Thank you very much!!

**Comment by papb** [comment]
Also, in my opinion, this issue should be reopened and considered a feature request.

**Comment by mglowinski93** [comment]
@aebrahim 
It's a nice solution:) 

Did you maybe try to make it work with `html` page with `static` files?

**Comment by JarroVGIT** [comment]
> Hello @aebrahim and others, what if the request we want to proxy is streaming a lot of data as _input_? The solution by @aebrahim waits for all the request input to arrive first, and then will end up sending it all at once to the target server, am I right? 

No, this is not correct. It provides an generator that is passed to the `StreamingResponse`. As it receives a chunk of data, it will pass that chunk of data through to your client.

**Comment by papb** [comment]
> > Hello @aebrahim and others, what if the request we want to proxy is streaming a lot of data as _input_? The solution by @aebrahim waits for all the request input to arrive first, and then will end up sending it all at once to the target server, am I right?
> 
> No, this is not correct. It provides an generator that is passed to the `StreamingResponse`. As it receives a chunk of data, it will pass that chunk of data through to your client.

@JarroVGIT are you sure? I think you're confusing input with output. I agree that their solution correctly streams the output of the target server (e.g. for downloading a large file from the target server). That's where the `StreamingResponse` comes in. However, I don't see anything that would do the same for the request (i.e. there's nothing like a `StreamingRequest` in their solution. Rather, what I see is an `await request.body()` which indicates what I said, that all the input is awaited before doing anything else.

**Comment by JarroVGIT** [comment]
```python
rp_req = client.build_request(request.method, url,
                                  headers=request.headers.raw,
                                  content=await request.body())
    rp_resp = await client.send(rp_req, stream=True)
```
The rp_resp is not yet fully loaded. Then, the streaming response is created with a generator, created from the `rp_resp` object (which starts the stream). 

```python
return StreamingResponse(
        rp_resp.aiter_raw(),
       ...
       )
```

I am not sure where you see `await request.body()`?

**Comment by morty29** [comment]
@papb I think [streaming requests on httpx doc](https://www.python-httpx.org/async/#streaming-requests) is what you are looking for. Meaning just remove the `await` from the assignment of content parameter in build request call in @aebrahim's answer and you are good

**Comment by papb** [comment]
@JarroVGIT `await request.body()` is definitely there, just look again maybe :)

@morty29 oh cool, thanks, I will try that.

**Comment by JarroVGIT** [comment]
> @JarroVGIT `await request.body()` is definitely there, just look again maybe :)
> 
> @morty29 oh cool, thanks, I will try that.

Whelp, you are right, my apologies!

**Comment by papb** [comment]
@JarroVGIT No problem :) Thank you for all the attention so far.

> @papb does the large request come in with uploaded files or with just a huge request body?

I don't have a specific request in mind, so please assume that both cases may happen. My goal is to understand how to make the best possible reverse-proxy-like API handler in FastAPI. The solution by @aebrahim covered all cases I could think of except for requests with nontrivial inputs (such as a huge request body), so that's why I asked about that.

By the way I saw you recommending [just using a real reverse proxy](https://github.com/tiangolo/fastapi/issues/5368#issuecomment-1242678186) in a comment for another issue, and for my use case I am currently doing just that (using [mitmproxy](https://mitmproxy.org/)), but I found it cumbersome to have something else running along FastAPI (since I need more things than just a reverse-proxy). I feel that the suggestion by @morty29 probably works, I just haven't had the time to try it out yet.

**Comment by DrecDroid** [comment]
To pass incoming request streaming data just pass `request.stream()` as `content=`


> ```python
> from starlette.requests import Request
> from starlette.responses import StreamingResponse
> from starlette.background import BackgroundTask
> 
> import httpx
> 
> client = httpx.AsyncClient(base_url="http://containername:7800/")
> 
> async def _reverse_proxy(request: Request):
>     url = httpx.URL(path=request.url.path,
>                     query=request.url.query.encode("utf-8"))
>     rp_req = client.build_request(request.method, url,
>                                   headers=request.headers.raw,
>                                   content=request.stream())
>     rp_resp = await client.send(rp_req, stream=True)
>     return StreamingResponse(
>         rp_resp.aiter_raw(),
>         status_code=rp_resp.status_code,
>         headers=rp_resp.headers,
>         background=BackgroundTask(rp_resp.aclose),
>     )
> 
> app.add_route("/titles/{path:path}",
>               _reverse_proxy, ["GET", "POST"])
> ```
